src/gui/styles_dialog.c

```diff
@@ -819,31 +819,45 @@ typedef struct _preview_data_t
 {
   char style_name[128];
   int imgid;
+  gboolean first_draw;
+  cairo_surface_t *surface;
 } _preview_data_t;
 
 static gboolean _preview_draw(GtkWidget *widget, cairo_t *cr, gpointer user_data)
 {
   _preview_data_t *data = (_preview_data_t *)user_data;
-  gboolean res = FALSE;
 
-  if(data->imgid > 0)
+  if(data->imgid > 0 && !data->first_draw && !data->surface)
+    data->surface = dt_gui_get_style_preview(data->imgid, data->style_name);
+
+  if(data->surface)
   {
-    cairo_surface_t *surface = dt_gui_get_style_preview(data->imgid, data->style_name);
     const int psize = dt_conf_get_int("ui/style/preview_size");
-    const int swidth = cairo_image_surface_get_width(surface);
-    const int sheight = cairo_image_surface_get_height(surface);
-    cairo_set_source_surface(cr, surface, .5f * (psize - swidth), .5f * (psize - sheight));
+    const int swidth = cairo_image_surface_get_width(data->surface);
+    const int sheight = cairo_image_surface_get_height(data->surface);
+    cairo_set_source_surface(cr, data->surface, .5f * (psize - swidth), .5f * (psize - sheight));
     cairo_paint(cr);
-    cairo_surface_destroy(surface);
-
-    res=TRUE;
+  }
+  else
+  {
+    data->first_draw = FALSE;
+    gtk_widget_queue_draw(widget);
   }
 
-  return res;
+  return FALSE;
 }
 
 GtkWidget *dt_gui_style_content_dialog(char *name, const int imgid)
 {
+  static _preview_data_t data = { "", -1, FALSE, NULL};
+
+  if(data.surface
+     && (strcmp(data.style_name, name) || imgid != data.imgid))
```

Comment:
I would also use the current history hash using `dt_history_hash_read` in `common/history.h`.

This will solve the issue with image being edited since the preview was computed.

Reply:
Alternatively we could get the image cache and use `change_timestamp` from `dt_image_t` struct but this will probably be slower.

Reply:
For the second point, after editing a style I would just clear the cache to ensure next run won't match. That is resetting the `style_name` in data struct above to "".